File: JavaScript/hw5/Vector.js
class Vector{
    constructor(v){
        if (v.length < 1)
            throw new Error('Dimension less than 1');
        this.dim = v.length;
        this.comp = v;
    }

    dimension(){
        return this.dim;
    }

    component(opts = {}){
        if(!opts['index'])
            return this.comp;
        return this.comp[opts['index']-1];
    }

    length(){
        var s = 0;
        for(var i in this.comp)
            s += this.comp[i] * this.comp[i];
        return Math.sqrt(s);
    }

    scalar(a) {
        var v = [];
        for (i in this.comp) {
            v.push(this.comp[i] * a);
        }
        return new Vector(v);
    }

    static sum (v1, v2) {
        if (v1.dimension() != v2.dimension())
            throw new Error('Dimension error');
        var v = [];
        for (var i in v1.component()) {
            v.push(v1.component()[i] + v2.component()[i]);
        }
        return new Vector(v);
    }

    static productDot(v1, v2) {
        if (v1.dimension() != v2.dimension())
            throw new Error('Dimension error');
        var s = 0;
        for (var i in v1.component()) {
            s += v1.component()[i] * v2.component()[i];
        }
        return s;
    }
}
function getVector(s){
    if((s.charAt(0) !== '(') || (s.charAt(s.length - 1) !== ')'))
        throw new Error('Format Error');
    s = s.substring(1, s.length - 1);
    s = s.split(',');
    v = [];
    for(i = 0; i < s.length; i++)
        v.push(parseInt(s[i]));
    return new Vector(v);
}

File: Python/hw5/Vector.py
import math

class Vector():
    def __init__(self, comp):
        if len(comp) <= 0:
            raise Exception('Dimension Error')
        self.dim = len(comp)
        self.comp = comp

    def dimension(self):
        return self.dim

    def component(self, index):
        return self.comp[index - 1]

    def __add__(self, v):
        if self.dim != v.dim:
            raise Exception('Dimension Error')
        i = 1
        l = []
        while i <= self.dim:
             l.append(self.component(i) + v.component(i))
             i+=1
        return Vector(l)

    def __sub__(self, v):
        if self.dim != v.dim:
            raise Exception('Dimension Error')
        i = 1
        l = []
        while i <= self.dim:
            l.append(self.component(i) - v.component(i))
            i += 1
        return Vector(l)


    def __mul__(self, v):
        t = v.__class__.__name__
        if t == 'Vector':
            return self.innerProduct(v)
        elif t == 'int':
            return self.productScalar(v)

    __rmul__ = __mul__
    
    def innerProduct(self, v):
        if self.dim != v.dim:
            raise Exception('Dimension Error')
        i = 0
        result = 0
        while i < self.dim:
            result += self.comp[i] * v.comp[i]
            i += 1
        return result

    def length(self):
        result = 0
        for i in self.comp:
            result += i ** 2
        return math.sqrt(result)

    def productScalar(self, a):
        l = []
        for i in self.comp:
            l.append(i * a)
            i += 1
        return Vector(l)
